src/libraries/System.Security.Cryptography.X509Certificates/tests/MatchesHostnameTests.cs

```diff
@@ -0,0 +1,758 @@
+                    AssertMatch(false, cert, ".");
+                    AssertMatch(true, cert, "*");
+                    AssertMatch(true, cert, "*.");
+                }
+            }
+        }
+
+        [Fact]
+        public static void TooManySANsThrows()
+        {
+            byte[] tooManySans = (
+                "3082021430820175A00302010202083C883E44C34DA5CB300A06082A8648CE3D" +
+                "04030230233121301F06035504031318706F7461746F2E766567657461626C65" +
+                "2E6578616D706C65301E170D3232303630383232333530365A170D3232303630" +
+                "383232333730365A30233121301F06035504031318706F7461746F2E76656765" +
+                "7461626C652E6578616D706C6530819B301006072A8648CE3D020106052B8104" +
+                "0023038186000400BA92930960C2C98D81F4DEAB62E75C0F768B5518A8FF58C2" +
+                "1D43B453AA2D1C73FA6BB0586349DDD61D0C25DC46B444BF5806F72F0F83546C" +
+                "B27583AE0007B101780007B7AE5717D4343C85D168212F2C2E4EC8F8B9F1953F" +
+                "A159C5E74A191B609E6A38FAAC404E3A0C094DD39A6732673545EE8C195A2B9B" +
+                "600420E9F55C145232304EA350304E30180603551D110411300F820D66727569" +
+                "742E6578616D706C6530320603551D11042B302982152A2E64727570652E6672" +
+                "7569742E6578616D706C65811069744066727569742E6578616D706C65300A06" +
+                "082A8648CE3D04030203818C003081880242009DA8DF6009D12EC733ADEE7479" +
+                "18B4611E185E478BA1D33AB7150A6A29F21FF31B48846B132868934A9F989C88" +
+                "39C7B8955A70DD5D4E9E1BB7C0D78F6AD8C3C6DC024200958482B9444D1AD2D3" +
+                "F67B51AD13064F2FDD4EC2F64ECB352D3F11BE8066F9021DD0CF309654351781" +
+                "69E940B767111BB2D28119EB3A2461617792F1CDF131F794").HexToByteArray();
+
+            using (X509Certificate2 cert = new X509Certificate2(tooManySans))
+            {
+                Assert.Throws<CryptographicException>(
+                    () => cert.MatchesHostname("fruit.example"));
+
+                Assert.Throws<CryptographicException>(
+                    () => cert.MatchesHostname("fruit.example", allowWildcards: false));
+
+                Assert.Throws<CryptographicException>(
+                    () => cert.MatchesHostname("fruit.example", allowCommonName: false));
+
+                Assert.Throws<CryptographicException>(
+                    () => cert.MatchesHostname("fruit.example", false, false));
+
+                // But argument validation comes first.
+                Assert.Throws<ArgumentNullException>("hostname", () => cert.MatchesHostname(null));
+                Assert.Throws<ArgumentNullException>("hostname", () => cert.MatchesHostname(null, false, true));
+                Assert.Throws<ArgumentNullException>("hostname", () => cert.MatchesHostname(null, true, false));
+                Assert.Throws<ArgumentNullException>("hostname", () => cert.MatchesHostname(null, false, false));
+            }
+        }
```

Comment:
Some other test scenarios I would be interested in:

1. Multiple wildcards in the SAN (yes, bogus, but people will try) `*.*.rambutan.example`
2. Things that are not host names in the `MatchesHostname`. What happens when someone inevitably puts user input in to this parameter?

If I am looking at this correctly, a certificate with a SAN of `*.potato.example` will, I think incorrectly, match `https://dog?example.potato.example`.

Reply:
Spending a bit of time noodling on this, this test fails:


```C#
[Fact]
public static void Bloop()
{
using (ECDsa key = ECDsa.Create())
{
    CertificateRequest req = new("CN=10.0.0.1", key, HashAlgorithmName.SHA256);
    SubjectAlternativeNameBuilder sanBuilder = new SubjectAlternativeNameBuilder();
    sanBuilder.AddDnsName("*.potato.example");

    req.CertificateExtensions.Add(sanBuilder.Build());

    DateTimeOffset now = DateTimeOffset.UtcNow;
    DateTimeOffset notBefore = now.AddMinutes(-1);
    DateTimeOffset notAfter = now.AddMinutes(1);

    using (X509Certificate2 cert = req.CreateSelfSigned(notBefore, notAfter))
    {
        AssertMatch(false, cert, "https://dog?example.potato.example");
    }
}
```

Reply:
Adding a precondition check hasn't made too many things fail... mainly just where there's no SAN being restrictive against what we match in CN.... which is probably OK:

```
  EEP? not a hostname (Unknown, '*.fruit.example') (Parameter 'hostname')
  EEP? not a hostname (Unknown, '*.fruit.example') (Parameter 'hostname')
  not a hostname (Unknown, '.') (Parameter 'hostname')
  EEP? not a hostname (Unknown, '*') (Parameter 'hostname')
  EEP? not a hostname (Unknown, '*.') (Parameter 'hostname')
  EEP? not a hostname (Unknown, '*.fruit.example') (Parameter 'hostname')
  EEP? not a hostname (Unknown, '*.fruit.example') (Parameter 'hostname')
  not a hostname (Unknown, '.') (Parameter 'hostname')
  not a hostname (Unknown, '*') (Parameter 'hostname')
  not a hostname (Unknown, '*.') (Parameter 'hostname')
  EEP? not a hostname (Unknown, '*berry.fruit.example') (Parameter 'hostname')
  EEP? not a hostname (Unknown, 'cran*.fruit.example') (Parameter 'hostname')
  not a hostname (Unknown, '.FRUit.example.') (Parameter 'hostname')
```

"EEP?" means the test was currently expecting it to return true.

I'm probably not too bothered by that.

Reply:
I'm not 100% sure how to parse that, but you seem unbothered by the output, so I'll assume it's a Good Thing™.